// yandex-dialogs-sdk/dist/imagesApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_fetch_1 = __importDefault(require("node-fetch"));
const constants_1 = require("./constants");
class ImagesApi {
    constructor(params) {
        this._skillId = params.skillId;
        this._oAuthToken = params.oAuthToken;
    }
    async uploadImageByUrl(url) {
        const response = await this._makeRequest({
            path: 'images',
            method: 'POST',
            body: { url },
        });
        return response.image;
    }
    async uploadImageFile() {
        throw new Error('Not implemented');
    }
    async getImages() {
        const response = await this._makeRequest({
            path: 'images',
            method: 'GET',
        });
        return response.images;
    }
    async _makeRequest(params) {
        const url = `${constants_1.ALICE_API_URL}/${this._skillId}/${params.path}`;
        const method = params.method || 'GET';
        const body = params.body ? JSON.stringify(params.body) : undefined;
        const response = await node_fetch_1.default(url, {
            method: method,
            headers: {
                Authorization: `OAuth ${this._oAuthToken}`,
                'Content-type': 'application/json',
            },
            body: body,
        });
        return await response.json();
    }
}
exports.ImagesApi = ImagesApi;
